**nota**: rispetto a 01 sono stati tolti i suspicious =1 e le impression =0, è stata stata rifatta parte di analisi utile nel report.

# Importazione dati

In [1]:
from google.colab import drive 
import pandas as pd
import re
from glob import glob
import numpy as np
from math import exp

from sklearn import svm
from sklearn.ensemble import RandomForestClassifier

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

from sklearn.metrics import classification_report
from sklearn.metrics import RocCurveDisplay
from sklearn.metrics import ConfusionMatrixDisplay

from imblearn.under_sampling import RandomUnderSampler
from collections import Counter
from imblearn.over_sampling import RandomOverSampler
from imblearn.over_sampling import SMOTENC

import scipy

In [2]:
drive.mount('/content/gdrive')

Mounted at /content/gdrive


Perché il seguente comando funzioni è necessario aggiungere una scorciatoia a DS Lab - PROGETTO al proprio Drive.

In [3]:
#!cp -av '/content/gdrive/My Drive/DS Lab - PROGETTO/Dataset aggregati' 'campaignClickDataset'

Lista dei file contenenti i dati

In [4]:
#part_files = sorted(glob('campaignClickDataset/part*.csv'))

In [5]:
#df = pd.concat((pd.read_csv(file) for file in part_files), ignore_index=True)

In [6]:
# Export in csv
#pd.DataFrame.to_csv(df.set_index('ad_form_id'), '/content/gdrive/My Drive/DS Lab - PROGETTO/Ottobre/dataset_completo.csv')

Importazione dei dati, solo csv completo

In [7]:
df = pd.read_csv('/content/gdrive/My Drive/DS Lab - PROGETTO/Ottobre/dataset_completo.csv')

In [8]:
df.head()

ad_form_id  suspicious  clicks  impressions  buy  os_android  \
0  7241135337730459930           0       0            1    0           1   
1  8408205703932483258           0       0            1    0           0   
2  5884145627470739741           0       0            2    0           1   
3  2463214613830580928           0       0            6    0           1   
4  6157419832685345406           0       0            3    0           0   

   os_bsd  os_ios  os_linux  os_osx  ...  feelings1_surprise  \
0       0       0         0       0  ...                 0.0   
1       0       0         0       0  ...                 0.0   
2       0       0         0       0  ...                 0.0   
3       0       0         0       0  ...                 0.0   
4       0       0         0       0  ...                 0.0   

   feelings1_sympathy  feelings1_thirst  feelings1_thoughtful  \
0                 0.0               0.0                   0.0   
1                 0.0               0.0                   0.0   
2                 0.0               0.0                   0.0   
3                 0.0               0.0                   0.0   
4                 0.0               0.0                   0.0   

   feelings1_torment  feelings1_traditionalist  feelings1_trust  \
0                0.0                       0.0              0.0   
1                0.0                       0.0              0.0   
2                0.0                       0.0              0.0   
3                0.0                       0.0              0.0   
4                0.0                       0.0              0.0   

   feelings1_uncategorized  feelings1_violence  feelings1_wealth  
0                      0.0                 0.0               0.0  
1                      0.0                 0.0               0.0  
2                      0.0                 0.0               0.0  
3                      0.0                 0.0               0.0  
4                      0.0                 0.0               0.0  

[5 rows x 1416 columns]

In [10]:
# Numero di sì
len(df[df['clicks'] != 0])

273

# Colonne
Idee per la rimozione delle colonne:
* soli 0
  * buy
  * sentiments
* identiche informazioni
  * time2, stessa informazione di time1
* correlazione + identiche informazioni
  * 57% degli admants correlate con categories1 -> dopo aver eliminato le categories che causano i nan nel calcolo della correlazione i risultati non sono cambiati
* uguaglianza delle distribuzioni + identiche informazioni
  * categories 1, 2 e 3 => rappresentano le stesse informazioni, ma le c1 sono più diverse a livello di distribuzione tra click e non click rispetto a c2 e c3
  * feelings => la quasi totalità delle colonne (93%) ha la stessa distribuzione tra click e non click (sono di fatto quasi tutti 0) -> la percentuale di 0 nelle colonne è minore del 99% solo in 4 casi: due colonne hanno % di 0 pari a circa il 98%, mentre le altre 2 hanno % di 0 minore del 77%. Queste ultime due colonne hanno anche distribuzione diversa tra click e non click e % di 0 diverse tra click e non click. Una possibile idea è quella di mantenere 3 colonne, queste due più una terza, feelings_others, in cui accorpiamo tutti gli altri feelings. Tuttavia, più di 44000 righe hanno tutti i feelings pari a 0 e questo sembra una mancanza di informazione importante e quindi si può anche valutare di eliminare tutte le colonne feelings.

In [ ]:
# Eliminazione con impressions = 0 (click passati da email) e suspicious = 1 (sospette frodi)

df  = df[df['impressions'] != 0]
df = df[df['suspicious'] == 0]

In [ ]:
# eliminazione colonna sus

df.drop(columns='suspicious', inplace=True)

In [ ]:
colonne = list(df.columns)
for colonna in colonne:
  print(colonna)

ad_form_id
clicks
impressions
buy
os_android
os_bsd
os_ios
os_linux
os_osx
os_other
os_windows
device_type
browser_android
browser_chrome
browser_chromium
browser_edge
browser_firefox
browser_ie
browser_opera
browser_other
browser_safari
browser_unknown
time1_workday_morning
time1_workday_afternoon
time1_workday_evening
time1_workday_night
time1_weekend_morning
time1_weekend_afternoon
time1_weekend_evening
time1_weekend_night
time2_morning_early
time2_morning
time2_launch
time2_afternoon
time2_evening
time2_night
time2_sleep
L00_50
L51_100
L101_250
L251_500
L501_1000
L1001_2500
L2501_5000
L5001_10000
L10001_more
categories1_artandentertainment
categories1_automotive
categories1_business
categories1_careers
categories1_education
categories1_emotions
categories1_familyandparenting
categories1_finance
categories1_foodanddrink
categories1_healthandfitness
categories1_hobbiesandinterests
categories1_homeandgarden
categories1_intentions
categories1_lawgovtandpolitics
categories1_news
categor

## Buy

Togliamo buy perché è totalmente popolata da 0 e eventuali valori pari a 1 indicherebbero che una persona ha cliccato dopo aver ricevuto la pubblicità => è un'informazione che non abbiamo prima di aver mandato la pubblicità.

In [ ]:
df.drop('buy', axis=1, inplace=True)

## Variabili Lunghezza
Per ora teniamo le colonne come sono, senza aggregarle. Poi eventualmente le aggreghiamo sulla base dei risultati di SVM.

### Normalizzazione

Normalizziamo le colonne (esattamente come abbiamo fatto a luglio).

Dall'analisi risulta che si sono circa 20000 righe con valori mancanti (nan o infiniti) per almeno una delle colonne. L'eliminazione la facciamo sulla base della somma (eliminiamo se è nan o infinito).

In [ ]:
L = re.compile('^L+.*', re.IGNORECASE)
lung = []

for v in df.columns:
  if L.search(v):
    print(v)
    lung.append(v)

L00_50
L51_100
L101_250
L251_500
L501_1000
L1001_2500
L2501_5000
L5001_10000
L10001_more


In [ ]:
# Esempio di ciò che succede coi NaN
df[df['L101_250'].isnull()][lung]

L00_50  L51_100  L101_250  L251_500  L501_1000  L1001_2500  L2501_5000  \
0         NaN      NaN       NaN       NaN        NaN         NaN         NaN   
2         inf      NaN       NaN       NaN        NaN         NaN         NaN   
3         inf      NaN       NaN       NaN        NaN         NaN         NaN   
6         inf      NaN       NaN       NaN        NaN         NaN         NaN   
11        inf      NaN       NaN       NaN        NaN         NaN         NaN   
...       ...      ...       ...       ...        ...         ...         ...   
82529     inf      NaN       NaN       NaN        NaN         NaN         NaN   
82532     NaN      NaN       NaN       NaN        NaN         NaN         NaN   
82534     inf      NaN       NaN       NaN        NaN         NaN         NaN   
82535     inf      NaN       NaN       NaN        NaN         NaN         NaN   
82553     inf      NaN       NaN       NaN        NaN         NaN         NaN   

       L5001_10000  L10001_more  
0              NaN          NaN  
2              NaN          NaN  
3              NaN          NaN  
6              NaN          NaN  
11             NaN          NaN  
...            ...          ...  
82529          NaN          NaN  
82532          NaN          NaN  
82534          NaN          NaN  
82535          NaN          NaN  
82553          NaN          NaN  

[20006 rows x 9 columns]

In [ ]:
# Somma dei valori nelle colonne
def somma_lung(riga):
  sum = 0
  for el in lung:
    sum += riga[el]
  return sum

df['somma'] = df.apply(somma_lung, axis=1)
df[['somma']]

somma
0        NaN
1      237.5
2        NaN
3        NaN
4      100.0
...      ...
82559  100.0
82560  100.0
82561  100.0
82562  100.0
82563  100.0

[82390 rows x 1 columns]

In [ ]:
# Verifichiamo se ci sono valori infiniti
from numpy import inf

sum(df['somma'] == inf) 

0

In [ ]:
# Verifichiamo se ci sono valori nulli
sum(df['somma'].isnull())

20011

Eliminiamo ora tutte queste righe.

In [ ]:
indexNames = df[(df['somma'].isnull())].index
df.drop(indexNames, inplace=True)
df['somma'].shape

(62379,)

Ora normalizziamo

In [ ]:
for l in lung:
  df[l] = round((df[l] / df['somma']) * 100, 2)

df['somma'] = df.apply(somma_lung, axis=1)

In [ ]:
# controllo
df[['L00_50', 'L51_100', 'L101_250', 'L251_500', 'L501_1000', 'L1001_2500', 'L2501_5000', 'L5001_10000', 'L10001_more', 'somma']]

L00_50  L51_100  L101_250  L251_500  L501_1000  L1001_2500  L2501_5000  \
1      100.00     0.00       0.0       0.0        0.0         0.0         0.0   
4      100.00     0.00       0.0       0.0        0.0         0.0         0.0   
5        0.00     0.00       0.0     100.0        0.0         0.0         0.0   
7       66.67    33.33       0.0       0.0        0.0         0.0         0.0   
8       42.86    57.14       0.0       0.0        0.0         0.0         0.0   
...       ...      ...       ...       ...        ...         ...         ...   
82559    0.00   100.00       0.0       0.0        0.0         0.0         0.0   
82560  100.00     0.00       0.0       0.0        0.0         0.0         0.0   
82561  100.00     0.00       0.0       0.0        0.0         0.0         0.0   
82562   11.11    88.89       0.0       0.0        0.0         0.0         0.0   
82563    0.00     0.00     100.0       0.0        0.0         0.0         0.0   

       L5001_10000  L10001_more  somma  
1              0.0          0.0  100.0  
4              0.0          0.0  100.0  
5              0.0          0.0  100.0  
7              0.0          0.0  100.0  
8              0.0          0.0  100.0  
...            ...          ...    ...  
82559          0.0          0.0  100.0  
82560          0.0          0.0  100.0  
82561          0.0          0.0  100.0  
82562          0.0          0.0  100.0  
82563          0.0          0.0  100.0  

[62379 rows x 10 columns]

In [ ]:
# Delete coloumn 'somma'
df.drop(columns=['somma'], inplace=True)

In [ ]:
# Controllo se ci sono colonne con valori pari a inf
for col in df.columns:
  if len(df[df[col] == np.inf]) != 0:
    print(col)

## Variabili os, browser e device

Tipi:
* os e browser sono composte da più variabili binarie;
* device è categorica, quindi si occuperà `OneHotEncoder` di binarizzarla.

Possiamo accorpare sulla base della loro distribuzione.

### os Distribuzione

In [ ]:
os = ('os_android', 'os_bsd', 'os_ios', 'os_linux', 'os_osx', 'os_other', 'os_windows')

In [ ]:
for tipo in os:
  yes = df[df['clicks'] > 0][tipo]
  no = df[df['clicks'] <= 0][tipo]

  zinyes = 0
  for r in yes:
    if r <= 0:
      zinyes = zinyes +1

  zinno = 0
  for r in no:
    if r <= 0:
      zinno = zinno +1

  print(tipo, ':')
  print('clicker che usano questo tipo: ', len(yes)-zinyes, 'su ', len(yes), ' -> ', 1-zinyes/len(yes))
  print('non clicker che usano questo tipo: ', len(no)-zinno, 'su ', len(no), ' -> ', 1-zinno/len(no))
  print('')

os_android :
clicker che usano questo tipo:  10 su  189  ->  0.05291005291005291
non clicker che usano questo tipo:  2191 su  62190  ->  0.03523074449268371

os_bsd :
clicker che usano questo tipo:  0 su  189  ->  0.0
non clicker che usano questo tipo:  1 su  62190  ->  1.607975558770569e-05

os_ios :
clicker che usano questo tipo:  0 su  189  ->  0.0
non clicker che usano questo tipo:  0 su  62190  ->  0.0

os_linux :
clicker che usano questo tipo:  2 su  189  ->  0.010582010582010581
non clicker che usano questo tipo:  461 su  62190  ->  0.007412767325936653

os_osx :
clicker che usano questo tipo:  24 su  189  ->  0.12698412698412698
non clicker che usano questo tipo:  9506 su  62190  ->  0.15285415661681945

os_other :
clicker che usano questo tipo:  0 su  189  ->  0.0
non clicker che usano questo tipo:  4 su  62190  ->  6.431902235082276e-05

os_windows :
clicker che usano questo tipo:  153 su  189  ->  0.8095238095238095
non clicker che usano questo tipo:  50027 su  62190  ->  0.

### browser Distribuzione

In [ ]:
brow = ('browser_android', 'browser_chrome', 'browser_chromium', 'browser_edge', 'browser_firefox', 'browser_ie', 'browser_opera', 
        'browser_other', 'browser_safari', 'browser_unknown')

In [ ]:
for tipo in brow:
  yes = df[df['clicks'] > 0][tipo]
  no = df[df['clicks'] <= 0][tipo]

  zinyes = 0
  for r in yes:
    if r <= 0:
      zinyes = zinyes +1

  zinno = 0
  for r in no:
    if r <= 0:
      zinno = zinno +1

  print(tipo, ':')
  print('clicker che usano questo tipo: ', len(yes)-zinyes, 'su ', len(yes), ' -> ', 1-zinyes/len(yes))
  print('non clicker che usano questo tipo: ', len(no)-zinno, 'su ', len(no), ' -> ', 1-zinno/len(no))
  print('')

browser_android :
clicker che usano questo tipo:  1 su  189  ->  0.005291005291005346
non clicker che usano questo tipo:  47 su  62190  ->  0.0007557485126226116

browser_chrome :
clicker che usano questo tipo:  109 su  189  ->  0.5767195767195767
non clicker che usano questo tipo:  40440 su  62190  ->  0.6502653159671973

browser_chromium :
clicker che usano questo tipo:  0 su  189  ->  0.0
non clicker che usano questo tipo:  0 su  62190  ->  0.0

browser_edge :
clicker che usano questo tipo:  45 su  189  ->  0.23809523809523814
non clicker che usano questo tipo:  9931 su  62190  ->  0.15968805274159836

browser_firefox :
clicker che usano questo tipo:  15 su  189  ->  0.07936507936507942
non clicker che usano questo tipo:  3492 su  62190  ->  0.056150506512301024

browser_ie :
clicker che usano questo tipo:  3 su  189  ->  0.015873015873015928
non clicker che usano questo tipo:  2739 su  62190  ->  0.04404245055475153

browser_opera :
clicker che usano questo tipo:  0 su  189  ->  0.

### device Distribuzione

In [ ]:
len(df[df['clicks']>0])

189

In [ ]:
for n in (1,2,3):
  users = len(df[df['device_type']==n])
  c_users = len( df[(df['device_type']==n) & (df['clicks']>0)] )
  nc_users = len( df[(df['device_type']==n) & (df['clicks']<=0)] )
  if(n==1): print('Mobile')
  if(n==2): print('Desktop and Laptop')
  if(n==3): print('Unknown/Others')
  print('clicker che usano questo device: ', c_users, 'su ', 210, ' -> ', c_users/210)
  print('non clicker che usano questo device: ', nc_users, 'su ', len(df)-210, ' -> ', nc_users/(len(df)-210))
  print('utilizzatori totali: ', users)
  print(' ')

Mobile
clicker che usano questo device:  10 su  210  ->  0.047619047619047616
non clicker che usano questo device:  2192 su  62169  ->  0.03525873023532629
utilizzatori totali:  2202
 
Desktop and Laptop
clicker che usano questo device:  179 su  210  ->  0.8523809523809524
non clicker che usano questo device:  59996 su  62169  ->  0.9650468883205456
utilizzatori totali:  60175
 
Unknown/Others
clicker che usano questo device:  0 su  210  ->  0.0
non clicker che usano questo device:  2 su  62169  ->  3.217037430230501e-05
utilizzatori totali:  2
 


## Variabili Time

Contenendo time1 e time2 informazioni simili (momento della giornata in cui la pagina è stata visitata), ma avendo time2 infromazioni mancanti per circa 3000 righe, si è deciso di mantenere le sole colonne di time1.

Viene quindi
* mostrato che le time2 hanno informazioni mancanti (somme a zero);
* mostrare che le time1 hanno tutte le informazioni.

Passi successivi:
* rimuovere effettivamente le Time2;
* normalizzare le Time1.


### Controlli

In [ ]:
# colonne da utilizzare
colonne = df.columns.to_numpy()
t = re.compile('^time+.*', re.IGNORECASE)
t1 = re.compile('^time1+.*', re.IGNORECASE)
times1 = []
times2 = []

for v in colonne:
  if t.search(v):
    print(v)
    if t1.search(v):
      times1.append(v)
    else:
      times2.append(v)

time1_workday_morning
time1_workday_afternoon
time1_workday_evening
time1_workday_night
time1_weekend_morning
time1_weekend_afternoon
time1_weekend_evening
time1_weekend_night
time2_morning_early
time2_morning
time2_launch
time2_afternoon
time2_evening
time2_night
time2_sleep


In [ ]:
# calcolo la somma dei tempi1 e tempi2

df['sommat1'] = (df['time1_workday_morning'] + df['time1_workday_afternoon'] + df['time1_workday_evening'] + df['time1_workday_night'] +
                 df['time1_weekend_morning'] + df['time1_weekend_afternoon'] + df['time1_weekend_evening'] + df['time1_weekend_night'])

df['sommat2'] = (df['time2_morning_early'] + df['time2_morning'] + df['time2_launch'] + df['time2_afternoon'] + df['time2_evening'] +
                 df['time2_night'] + df['time2_sleep'])

In [ ]:
# non sono presenti somme infinite
len(df[(df['sommat1'] > 100000) | (df['sommat2'] > 100000)][['sommat1', 'sommat2']])

0

In [ ]:
# sono presenti più di 3000 righe che mostrano somma per tempi2 pari a zero (13 hanno cliccato)
print(len(df[df['sommat2'] == 0][['sommat1', 'sommat2']]))
print(len(df[(df['sommat2'] == 0) & (df['clicks'] > 0)][['sommat1', 'sommat2']]))

3424
10


In [ ]:
#  non sono presenti righe con somma a zero per tempi1
df[(df['sommat1'] == 0)][['sommat1', 'sommat2']]

Empty DataFrame
Columns: [sommat1, sommat2]
Index: []

In [ ]:
# non sono presenti valori nulli
print(len(df[df['sommat1'].isnull()]))
print(len(df[df['sommat2'].isnull()]))

0
0


### Normalizzazione time1, eliminazione time2

In [ ]:
# Riscalo i tempi(1) in modo che somma sia 100 
for t2 in times1:
  df[t2] = round(df[t2]/ df['sommat1']*100, 2)

times1.append('sommat1')
df[times1].head(5)

time1_workday_morning  time1_workday_afternoon  time1_workday_evening  \
1                  26.32                    73.68                   0.00   
4                   0.00                     0.00                   0.00   
5                   0.00                   100.00                   0.00   
7                   0.00                   100.00                   0.00   
8                  14.29                    14.29                  71.43   

   time1_workday_night  time1_weekend_morning  time1_weekend_afternoon  \
1                  0.0                    0.0                      0.0   
4                100.0                    0.0                      0.0   
5                  0.0                    0.0                      0.0   
7                  0.0                    0.0                      0.0   
8                  0.0                    0.0                      0.0   

   time1_weekend_evening  time1_weekend_night  sommat1  
1                    0.0                  0.0    237.5  
4                    0.0                  0.0    100.0  
5                    0.0                  0.0    100.0  
7                    0.0                  0.0    100.0  
8                    0.0                  0.0    175.0

In [ ]:
# elimino colonne di time2 e delle somme
df.drop(columns=['sommat1', 'sommat2'], inplace=True)
df.drop(columns=times2, inplace=True)

## Variabili Categories

Normalizziamo le colonne (come abbiamo fatto a luglio, includendo però anche le categories 2 e 3).

### Normalizzazione

#### Categories1

In [ ]:
# Colonne categories1
cat1 = re.compile('categories1_+.*', re.IGNORECASE)
cats1 = []
for label in colonne:
  occ = cat1.findall(label)
  if occ:
    cats1.append(occ[0])
  
# Calcolo somma
def somma_cat1(riga):
  sum = 0
  for cat in cats1:
    sum += riga[cat]
  return(round(sum, 1))

df['somma1'] = df.apply(somma_cat1, axis=1)

In [ ]:
# controllo quanti nan ci sono nelle somme: non ci sono nan

df[df['somma1'].isnull()][['ad_form_id', 'somma1']]

Empty DataFrame
Columns: [ad_form_id, somma1]
Index: []

In [ ]:
# controllo se ci sono nan nelle categorie singole: non ci sono

for v in cats1:
  nn = len(df[df[v].isnull()])
  if (nn != 0):
    print(v, nn)

In [ ]:
# controllo quanti inf ci sono nelle somme: non ci sono inf
df[df['somma1'] == np.inf][['ad_form_id', 'somma1']]

Empty DataFrame
Columns: [ad_form_id, somma1]
Index: []

Le righe che davano somme infinite nelle lunghezze, sono le stesse che a luglio davano somme infinite nelle categories.

In [ ]:
# controllo gli zero: non ce ne sono
df[df['somma1'] == 0][['ad_form_id', 'somma1']]

Empty DataFrame
Columns: [ad_form_id, somma1]
Index: []

In [ ]:
# Normalizzo le somme a 100
for cat in cats1:
  df[cat] = round((df[cat] / df['somma1']) * 100, 2)

In [ ]:
df[cats1]

categories1_artandentertainment  categories1_automotive  \
1                                  0.0                     0.0   
4                                100.0                     0.0   
5                                  0.0                     0.0   
7                                  0.0                     0.0   
8                                  0.0                     0.0   
...                                ...                     ...   
82559                              0.0                     0.0   
82560                            100.0                     0.0   
82561                            100.0                     0.0   
82562                              0.0                     0.0   
82563                              0.0                     0.0   

       categories1_business  categories1_careers  categories1_education  \
1                       0.0                42.11                    0.0   
4                       0.0                 0.00                    0.0   
5                       0.0                 0.00                    0.0   
7                       0.0                 0.00                    0.0   
8                       0.0                 0.00                    0.0   
...                     ...                  ...                    ...   
82559                   0.0                 0.00                    0.0   
82560                   0.0                 0.00                    0.0   
82561                   0.0                 0.00                    0.0   
82562                   0.0                 0.00                    0.0   
82563                   0.0                 0.00                    0.0   

       categories1_emotions  categories1_familyandparenting  \
1                       0.0                             0.0   
4                       0.0                             0.0   
5                       0.0                             0.0   
7                       0.0                             0.0   
8                       0.0                             0.0   
...                     ...                             ...   
82559                   0.0                             0.0   
82560                   0.0                             0.0   
82561                   0.0                             0.0   
82562                   0.0                             0.0   
82563                   0.0                             0.0   

       categories1_finance  categories1_foodanddrink  \
1                      0.0                       0.0   
4                      0.0                       0.0   
5                      0.0                       0.0   
7                      0.0                       0.0   
8                      0.0                       0.0   
...                    ...                       ...   
82559                  0.0                       0.0   
82560                  0.0                       0.0   
82561                  0.0                       0.0   
82562                  0.0                       0.0   
82563                  0.0                       0.0   

       categories1_healthandfitness  ...  categories1_realestate  \
1                               0.0  ...                     0.0   
4                               0.0  ...                     0.0   
5                               0.0  ...                     0.0   
7                               0.0  ...                     0.0   
8                               0.0  ...                     0.0   
...                             ...  ...                     ...   
82559                           0.0  ...                     0.0   
82560                           0.0  ...                     0.0   
82561                           0.0  ...                     0.0   
82562                           0.0  ...                     0.0   
82563                           0.0  ...                     0.0   

       categories1_religionandspirituality  categories1_science  \

In [ ]:
df.drop('somma1', axis=1, inplace=True)

#### Categories 2

In [ ]:
# Colonne categories2
cat2 = re.compile('categories2_+.*', re.IGNORECASE)
cats2 = []
for label in colonne:
  occ = cat2.findall(label)
  if occ:
    cats2.append(occ[0])
  
# Calcolo somma
def somma_cat2(riga):
  sum = 0
  for cat in cats2:
    sum += riga[cat]
  return(round(sum, 1))

df['somma2'] = df.apply(somma_cat2, axis=1)

In [ ]:
# controllo quanti nan ci sono nelle somme: non ci sono nan

df[df['somma2'].isnull()][['ad_form_id', 'somma2']]

Empty DataFrame
Columns: [ad_form_id, somma2]
Index: []

In [ ]:
# controllo se ci sono nan nelle categorie singole: non ci sono

for v in cats2:
  nn = len(df[df[v].isnull()])
  if (nn != 0):
    print(v, nn)

In [ ]:
# controllo quanti inf ci sono nelle somme: non ci sono inf
df[df['somma2'] == np.inf][['ad_form_id', 'somma2']]

Empty DataFrame
Columns: [ad_form_id, somma2]
Index: []

Le righe che davano somme infinite nelle lunghezze, sono le stesse che a luglio davano somme infinite nelle categories.

In [ ]:
# controllo gli zero: non ce ne sono
df[df['somma2'] == 0][['ad_form_id', 'somma2']]

Empty DataFrame
Columns: [ad_form_id, somma2]
Index: []

In [ ]:
# Normalizzo le somme a 100
for cat in cats2:
  df[cat] = round((df[cat] / df['somma2']) * 100, 2)

In [ ]:
df[cats2]

categories2_accessories  categories2_addiction  categories2_adoption  \
1                          0.0                    0.0                   0.0   
4                          0.0                    0.0                   0.0   
5                          0.0                    0.0                   0.0   
7                          0.0                    0.0                   0.0   
8                          0.0                    0.0                   0.0   
...                        ...                    ...                   ...   
82559                      0.0                    0.0                   0.0   
82560                      0.0                    0.0                   0.0   
82561                      0.0                    0.0                   0.0   
82562                      0.0                    0.0                   0.0   
82563                      0.0                    0.0                   0.0   

       categories2_adulteducation  categories2_adventuretravel  \
1                             0.0                          0.0   
4                             0.0                          0.0   
5                             0.0                          0.0   
7                             0.0                          0.0   
8                             0.0                          0.0   
...                           ...                          ...   
82559                         0.0                          0.0   
82560                         0.0                          0.0   
82561                         0.0                          0.0   
82562                         0.0                          0.0   
82563                         0.0                          0.0   

       categories2_advertising  categories2_agriculture  \
1                          0.0                      0.0   
4                          0.0                      0.0   
5                          0.0                      0.0   
7                          0.0                      0.0   
8                          0.0                      0.0   
...                        ...                      ...   
82559                      0.0                      0.0   
82560                      0.0                      0.0   
82561                      0.0                      0.0   
82562                      0.0                      0.0   
82563                      0.0                      0.0   

       categories2_alternativemedicine  categories2_alternativereligions  \
1                                  0.0                               0.0   
4                                  0.0                               0.0   
5                                  0.0                               0.0   
7                                  0.0                               0.0   
8                                  0.0                               0.0   
...                                ...                               ...   
82559                              0.0                               0.0   
82560                              0.0                               0.0   
82561                              0.0                               0.0   
82562                              0.0                               0.0   
82563                              0.0                               0.0   

       categories2_americancuisine  ...  categories2_waterskiwakeboard  \
1                              0.0  ...                            0.0   
4                              0.0  ...                            0.0   
5                              0.0  ...                            0.0   
7                              0.0  ...                            0.0   
8                              0.0  ...                            0.0   
...                            ...  ...                            ...   
82559                          0.0  ...                            0.0   
82560                          0.0  ...            

In [ ]:
df.drop('somma2', axis=1, inplace=True)

#### Categories 3

In [ ]:
# Colonne categories3
cat3 = re.compile('categories3_+.*', re.IGNORECASE)
cats3 = []
for label in colonne:
  occ = cat3.findall(label)
  if occ:
    cats3.append(occ[0])
  
# Calcolo somma
def somma_cat3(riga):
  sum = 0
  for cat in cats3:
    sum += riga[cat]
  return(round(sum, 1))

df['somma3'] = df.apply(somma_cat3, axis=1)

In [ ]:
# controllo quanti nan ci sono nelle somme: non ci sono nan

df[df['somma3'].isnull()][['ad_form_id', 'somma3']]

Empty DataFrame
Columns: [ad_form_id, somma3]
Index: []

In [ ]:
# controllo se ci sono nan nelle categorie singole: non ci sono

for v in cats3:
  nn = len(df[df[v].isnull()])
  if (nn != 0):
    print(v, nn)

In [ ]:
# controllo quanti inf ci sono nelle somme: non ci sono inf
df[df['somma3'] == np.inf][['ad_form_id', 'somma3']]

Empty DataFrame
Columns: [ad_form_id, somma3]
Index: []

Le righe che davano somme infinite nelle lunghezze, sono le stesse che a luglio davano somme infinite nelle categories.

In [ ]:
# controllo gli zero: non ce ne sono
df[df['somma3'] == 0][['ad_form_id', 'somma3']]

Empty DataFrame
Columns: [ad_form_id, somma3]
Index: []

In [ ]:
# Normalizzo le somme a 100
for cat in cats3:
  df[cat] = round((df[cat] / df['somma3']) * 100, 2)

In [ ]:
df[cats3]

categories3_712education  categories3_accessories  \
1                           0.0                      0.0   
4                           0.0                      0.0   
5                           0.0                      0.0   
7                           0.0                      0.0   
8                           0.0                      0.0   
...                         ...                      ...   
82559                       0.0                      0.0   
82560                       0.0                      0.0   
82561                       0.0                      0.0   
82562                       0.0                      0.0   
82563                       0.0                      0.0   

       categories3_acoustics  categories3_active  categories3_addiction  \
1                        0.0                 0.0                    0.0   
4                        0.0                 0.0                    0.0   
5                        0.0                 0.0                    0.0   
7                        0.0                 0.0                    0.0   
8                        0.0                 0.0                    0.0   
...                      ...                 ...                    ...   
82559                    0.0                 0.0                    0.0   
82560                    0.0                 0.0                    0.0   
82561                    0.0                 0.0                    0.0   
82562                    0.0                 0.0                    0.0   
82563                    0.0                 0.0                    0.0   

       categories3_adoption  categories3_adulteducation  \
1                       0.0                         0.0   
4                       0.0                         0.0   
5                       0.0                         0.0   
7                       0.0                         0.0   
8                       0.0                         0.0   
...                     ...                         ...   
82559                   0.0                         0.0   
82560                   0.0                         0.0   
82561                   0.0                         0.0   
82562                   0.0                         0.0   
82563                   0.0                         0.0   

       categories3_adventuretravel  categories3_advertising  \
1                              0.0                      0.0   
4                              0.0                      0.0   
5                              0.0                      0.0   
7                              0.0                      0.0   
8                              0.0                      0.0   
...                            ...                      ...   
82559                          0.0                      0.0   
82560                          0.0                      0.0   
82561                          0.0                      0.0   
82562                          0.0                      0.0   
82563                          0.0                      0.0   

       categories3_affection  ...  categories3_weightloss  \
1                        0.0  ...                     0.0   
4                        0.0  ...                     0.0   
5                        0.0  ...                     0.0   
7                        0.0  ...                     0.0   
8                        0.0  ...                     0.0   
...                      ...  ...                     ...   
82559                    0.0  ...                     0.0   
82560                    0.0  ...                     0.0   
82561                    0.0  ...                     0.0   
82562                    0.0  ...                     0.0   
82563                    0.0  ...                     0.0   

       categories3_welfare  categories3_windows  categories3_windsurfing  \
1                      0.0                  0.0                      0.0   
4                      0.0        

In [ ]:
df.drop('somma3', axis=1, inplace=True)

### Distribuzione

https://machinelearningmastery.com/statistical-hypothesis-tests-in-python-cheat-sheet/

In [ ]:
df_y = df[df['clicks'] != 0]
df_n = df[df['clicks'] == 0]

In [ ]:
alfa=0.05

#### Categories 1

In [ ]:
dist_c1 = [] # vettore di dizionari in cui salvo il p-value del test di student
conteggio_distr_identiche_c1 = 0
for cat in cats1:
    diz = {}

    _, p = scipy.stats.mannwhitneyu(df_y[cat], df_n[cat])

    diz['C1'] = cat
    diz['p-value'] = p

    if diz['p-value'] > alfa:
        diz['Esito'] = 'Distribuzione identica'
        conteggio_distr_identiche_c1 += 1
    
    else:
        diz['Esito'] = 'Distribuzione diversa'
    
    dist_c1.append(diz)

#### Categories 2

In [ ]:
dist_c2 = [] # vettore di dizionari in cui salvo il p-value del test di student
conteggio_distr_identiche_c2 = 0
for cat in cats2:
    diz = {}

    _, p = scipy.stats.mannwhitneyu(df_y[cat], df_n[cat])

    diz['C2'] = cat
    diz['p-value'] = p

    if diz['p-value'] > alfa:
        diz['Esito'] = 'Distribuzione identica'
        conteggio_distr_identiche_c2 += 1
    
    else:
        diz['Esito'] = 'Distribuzione diversa'
    
    dist_c2.append(diz)

#### Categories 3

In [ ]:
dist_c3 = [] # vettore di dizionari in cui salvo il p-value del test di student
conteggio_distr_identiche_c3 = 0
for cat in cats3:
    diz = {}

    _, p = scipy.stats.mannwhitneyu(df_y[cat], df_n[cat])

    diz['C3'] = cat
    diz['p-value'] = p

    if diz['p-value'] > alfa:
        diz['Esito'] = 'Distribuzione identica'
        conteggio_distr_identiche_c3 += 1
    
    else:
        diz['Esito'] = 'Distribuzione diversa'
    
    dist_c3.append(diz)

#### Confronto

In [ ]:
print(f"Frazione di Categories 1 con distribuzione identica tra click e non: {round(conteggio_distr_identiche_c1 / len(cats1), 2)}")
print(f"Frazione di Categories 2 con distribuzione identica tra click e non: {round(conteggio_distr_identiche_c2 / len(cats2), 2)}")
print(f"Frazione di Categories 3 con distribuzione identica tra click e non: {round(conteggio_distr_identiche_c3 / len(cats3), 2)}")

Frazione di Categories 1 con distribuzione identica tra click e non: 0.85
Frazione di Categories 2 con distribuzione identica tra click e non: 0.96
Frazione di Categories 3 con distribuzione identica tra click e non: 0.97


#### Analisi della frazione di 0 delle categories 1

##### Click e non click separati

In [ ]:
zeri_cnc = []
for cat in cats1:
  diz = {}
  zeri_y = len(df_y[df_y[cat] == 0])
  zeri_n = len(df_n[df_n[cat] == 0])

  diz['Categories1'] = cat
  diz['% 0 tra i click'] = round(zeri_y / len(df_y), 3)
  diz['% 0 tra i non click'] = round(zeri_n / len(df_n), 3)

  zeri_cnc.append(diz)

In [ ]:
for diz in zeri_cnc:
  print(diz)

{'Categories1': 'categories1_artandentertainment', '% 0 tra i click': 0.741, '% 0 tra i non click': 0.631}
{'Categories1': 'categories1_automotive', '% 0 tra i click': 0.995, '% 0 tra i non click': 0.99}
{'Categories1': 'categories1_business', '% 0 tra i click': 0.989, '% 0 tra i non click': 0.972}
{'Categories1': 'categories1_careers', '% 0 tra i click': 0.984, '% 0 tra i non click': 0.976}
{'Categories1': 'categories1_education', '% 0 tra i click': 0.995, '% 0 tra i non click': 0.99}
{'Categories1': 'categories1_emotions', '% 0 tra i click': 1.0, '% 0 tra i non click': 1.0}
{'Categories1': 'categories1_familyandparenting', '% 0 tra i click': 0.995, '% 0 tra i non click': 0.99}
{'Categories1': 'categories1_finance', '% 0 tra i click': 1.0, '% 0 tra i non click': 0.987}
{'Categories1': 'categories1_foodanddrink', '% 0 tra i click': 0.984, '% 0 tra i non click': 0.988}
{'Categories1': 'categories1_healthandfitness', '% 0 tra i click': 0.995, '% 0 tra i non click': 0.993}
{'Categories1':

Verifico se ci sono colonne con % di 0 minore di 0.99.

In [ ]:
index = 0 # conto quante categories rispettano la condizione
for diz in zeri_cnc:
  if diz['% 0 tra i click'] < 0.99 or diz['% 0 tra i non click'] < 0.99:
    index += 1
    print(diz)

{'Categories1': 'categories1_artandentertainment', '% 0 tra i click': 0.741, '% 0 tra i non click': 0.631}
{'Categories1': 'categories1_business', '% 0 tra i click': 0.989, '% 0 tra i non click': 0.972}
{'Categories1': 'categories1_careers', '% 0 tra i click': 0.984, '% 0 tra i non click': 0.976}
{'Categories1': 'categories1_finance', '% 0 tra i click': 1.0, '% 0 tra i non click': 0.987}
{'Categories1': 'categories1_foodanddrink', '% 0 tra i click': 0.984, '% 0 tra i non click': 0.988}
{'Categories1': 'categories1_hobbiesandinterests', '% 0 tra i click': 0.444, '% 0 tra i non click': 0.69}
{'Categories1': 'categories1_lawgovtandpolitics', '% 0 tra i click': 0.989, '% 0 tra i non click': 0.99}
{'Categories1': 'categories1_news', '% 0 tra i click': 0.984, '% 0 tra i non click': 0.988}
{'Categories1': 'categories1_pets', '% 0 tra i click': 0.984, '% 0 tra i non click': 0.983}
{'Categories1': 'categories1_science', '% 0 tra i click': 0.958, '% 0 tra i non click': 0.909}
{'Categories1': 'ca

In [ ]:
print(f"Numero categories1 con % zeri < 0.99 tra i click o tra i non click: {index}")
print(f"Totale categories1 {len(cats1)}")
print(f"Percentuale: {round(index / len(cats1), 3) * 100}%")

Numero categories1 con % zeri < 0.99 tra i click o tra i non click: 16
Totale categories1 26
Percentuale: 61.5%


##### Click e non click uniti

In [ ]:
zeri_tot = []
for cat in cats1:
  diz = {}
  zeri = len(df[df[cat] == 0])

  diz['Categories1'] = cat
  diz['% 0'] = round(zeri / len(df), 3)

  zeri_tot.append(diz)

In [ ]:
for diz in zeri_tot:
  print(diz)

{'Categories1': 'categories1_artandentertainment', '% 0': 0.632}
{'Categories1': 'categories1_automotive', '% 0': 0.99}
{'Categories1': 'categories1_business', '% 0': 0.973}
{'Categories1': 'categories1_careers', '% 0': 0.976}
{'Categories1': 'categories1_education', '% 0': 0.99}
{'Categories1': 'categories1_emotions', '% 0': 1.0}
{'Categories1': 'categories1_familyandparenting', '% 0': 0.99}
{'Categories1': 'categories1_finance', '% 0': 0.987}
{'Categories1': 'categories1_foodanddrink', '% 0': 0.988}
{'Categories1': 'categories1_healthandfitness', '% 0': 0.993}
{'Categories1': 'categories1_hobbiesandinterests', '% 0': 0.69}
{'Categories1': 'categories1_homeandgarden', '% 0': 0.994}
{'Categories1': 'categories1_intentions', '% 0': 0.993}
{'Categories1': 'categories1_lawgovtandpolitics', '% 0': 0.99}
{'Categories1': 'categories1_news', '% 0': 0.988}
{'Categories1': 'categories1_pets', '% 0': 0.983}
{'Categories1': 'categories1_realestate', '% 0': 0.998}
{'Categories1': 'categories1_reli

Verifico se ci sono colonne con % di 0 minore di 0.99.

In [ ]:
index = 0 # uso per contare le categories che rispettano la condizione qui sopra
for diz in zeri_tot:
  if diz['% 0'] < 0.99:
    index += 1
    print(diz)

{'Categories1': 'categories1_artandentertainment', '% 0': 0.632}
{'Categories1': 'categories1_business', '% 0': 0.973}
{'Categories1': 'categories1_careers', '% 0': 0.976}
{'Categories1': 'categories1_finance', '% 0': 0.987}
{'Categories1': 'categories1_foodanddrink', '% 0': 0.988}
{'Categories1': 'categories1_hobbiesandinterests', '% 0': 0.69}
{'Categories1': 'categories1_news', '% 0': 0.988}
{'Categories1': 'categories1_pets', '% 0': 0.983}
{'Categories1': 'categories1_science', '% 0': 0.909}
{'Categories1': 'categories1_society', '% 0': 0.989}
{'Categories1': 'categories1_sports', '% 0': 0.817}
{'Categories1': 'categories1_styleandfashion', '% 0': 0.988}
{'Categories1': 'categories1_technologyandcomputing', '% 0': 0.562}
{'Categories1': 'categories1_travel', '% 0': 0.977}
{'Categories1': 'categories1_uncategorized', '% 0': 0.874}


In [ ]:
print(f"Numero categories1 con % zeri < 0.99 tra i click o tra i non click: {index}")
print(f"Totale categories1 {len(cats1)}")
print(f"Percentuale: {round(index / len(cats1), 3) * 100}%")

Numero categories1 con % zeri < 0.99 tra i click o tra i non click: 15
Totale categories1 26
Percentuale: 57.699999999999996%


### Eliminazione Categories3

In [ ]:
df.drop(columns=cats3, inplace=True)

## Admants

### Normalizzazione

In [ ]:
adm = re.compile('admants+.*', re.IGNORECASE)
adms = []

for v in df.columns:
  if adm.search(v):
    adms.append(v)
  
# Calcolo somma
def somma_adms(riga):
  sum = 0
  for cat in adms:
    sum += riga[cat]
  return(round(sum, 1))

df['somma'] = df.apply(somma_adms, axis=1)

In [ ]:
# controllo quanti nan ci sono nelle somme: non ci sono nan

df[df['somma'].isnull()][['ad_form_id', 'somma']]

Empty DataFrame
Columns: [ad_form_id, somma]
Index: []

In [ ]:
# controllo se ci sono nan nelle categorie singole: non ci sono

for v in adms:
  nn = len(df[df[v].isnull()])
  if (nn != 0):
    print(v, nn)

In [ ]:
# controllo quanti inf ci sono nelle somme: non ci sono inf
df[df['somma'] == np.inf][['ad_form_id', 'somma']]

Empty DataFrame
Columns: [ad_form_id, somma]
Index: []

In [ ]:
# controllo gli zero: ce ne sono
df[df['somma'] == 0][['ad_form_id', 'somma']]

ad_form_id  somma
9      1520530489999888701    0.0
130     423758730060024042    0.0
176    2949209777750458222    0.0
236    8591797067637818061    0.0
248    8220246112129877727    0.0
...                    ...    ...
82250   806966730273731052    0.0
82254  7207700329399635762    0.0
82413  5081029587588027566    0.0
82431  5407638255918714062    0.0
82519  5731786672611993747    0.0

[2233 rows x 2 columns]

Guardiamo come sono i click e se ce ne sono di superiori a zero.

In [ ]:
df[(df['somma'] == 0) & (df['clicks'] != 0)][['ad_form_id', 'clicks', 'somma']]

ad_form_id  clicks  somma
10459  4441976255661567592       1    0.0
22712   824547121173218951       1    0.0
28247  4103523210449571236       1    0.0
38850  1119000281687526928       1    0.0
40988  2813289715831986827       1    0.0
47282   828253886724037402       1    0.0
49613  8703457363297317795       1    0.0
57979  4802750125375816531       1    0.0
71105  5019332339658836935       1    0.0
74386   505281467491500929       1    0.0
80411    60933584475798861       1    0.0

Per evitare di avere normalizzazioni a NaN: poniamo gli zero nelle somme uguali a un valore arbitrario in modo che la divisione non dia NaN.

In [ ]:
soglia = 10

def somma_soglia(riga):
  if riga['somma'] == 0:
    return soglia
  else:
    return riga['somma']
  
df['somma'] = df.apply(somma_soglia, axis=1)

In [ ]:
# Normalizzo le somme a 100
for cat in adms:
  df[cat] = round((df[cat] / df['somma']) * 100, 2)

In [ ]:
df[adms]

admants1_appliances  admants1_artandentertainment  admants1_automotive  \
1                      0.0                           0.0                  0.0   
4                      0.0                         100.0                  0.0   
5                      0.0                           0.0                  0.0   
7                      0.0                           0.0                  0.0   
8                      0.0                           0.0                  0.0   
...                    ...                           ...                  ...   
82559                  0.0                           0.0                  0.0   
82560                  0.0                         100.0                  0.0   
82561                  0.0                         100.0                  0.0   
82562                  0.0                           0.0                  0.0   
82563                  0.0                           0.0                  0.0   

       admants1_boatinsurance  admants1_businessandindustrialsupplies  \
1                         0.0                                     0.0   
4                         0.0                                     0.0   
5                         0.0                                     0.0   
7                         0.0                                     0.0   
8                         0.0                                     0.0   
...                       ...                                     ...   
82559                     0.0                                     0.0   
82560                     0.0                                     0.0   
82561                     0.0                                     0.0   
82562                     0.0                                     0.0   
82563                     0.0                                     0.0   

       admants1_careers  admants1_carinsurance  admants1_consumerelectronics  \
1                  50.0                    0.0                          0.00   
4                   0.0                    0.0                          0.00   
5                   0.0                    0.0                          0.00   
7                   0.0                    0.0                          0.00   
8                   0.0                    0.0                          0.00   
...                 ...                    ...                           ...   
82559               0.0                    0.0                          0.00   
82560               0.0                    0.0                          0.00   
82561               0.0                    0.0                          0.00   
82562               0.0                    0.0                          0.00   
82563               0.0                    0.0                         60.78   

       admants1_dating  admants1_education  ...  admants1_smartphones  \
1                  0.0                 0.0  ...                   0.0   
4                  0.0                 0.0  ...                   0.0   
5                  0.0                 0.0  ...                   0.0   
7                  0.0                 0.0  ...                   0.0   
8                  0.0                 0.0  ...                   0.0   
...                ...                 ...  ...                   ...   
82559              0.0                 0.0  ...                   0.0   
82560              0.0                 0.0  ...                   0.0   
82561              0.0                 0.0  ...                   0.0   
82562              0.0                 0.0  ...                   0.0   
82563              0.0                 0.0  ...                   0.0   

       admants1_software  admants1_sportinggoods  admants1_sportstickets  \
1                    0.0                    0.00                    0.00   
4                    0.0                    0.00                    0.00   
5                    0.0                    6.86                    6.86   
7              

In [ ]:
df.drop('somma', axis=1, inplace=True)

### Correlazione

Guardiamo ora la correlazione con le categories 1 in modo da decretare la fine delle une o delle altre (prof. diceva che le admants sono ottenute in modo automatizzato).

Utilizzo lo stesso procedimento usato per valutare la correlazione tra le categories.

In [ ]:
soglia_corr= 0.7

In [ ]:
# Punto 1
c1a = []
for adm in adms:
  for cat1 in cats1:
    if len(df[df[cat1] != 0]) > 0: # a causare i nan sono le categories con soli 0
      diz = {}
      diz['Adm'] = adm
      diz['C1'] = cat1
      diz['Corr'] = np.corrcoef(df[adm], df[cat1])[0][1]
      c1a.append(diz)

In [ ]:
# Punto 2
df1a = pd.DataFrame(c1a)
df1a.head()

Adm                               C1      Corr
0  admants1_appliances  categories1_artandentertainment -0.013822
1  admants1_appliances           categories1_automotive  0.001931
2  admants1_appliances             categories1_business  0.007459
3  admants1_appliances              categories1_careers -0.000602
4  admants1_appliances            categories1_education  0.000735

In [ ]:
df1a[df1a['Corr'].isnull()]

Empty DataFrame
Columns: [Adm, C1, Corr]
Index: []

In [ ]:
# Punto 3
c1a = []
for adm in adms:
    diz = {}
    diz['Adm'] = adm
    diz['C1'] = list(df1a[df1a['Adm'] == adm].sort_values('Corr', ascending=False)['C1'])[0]
    diz['Corr'] = list(df1a[df1a['Adm'] == adm].sort_values('Corr', ascending=False)['Corr'])[0]
    c1a.append(diz)

In [ ]:
# Punto 4
index = 0 # uso per valutare le correlazioni significative
for diz in c1a:
    print(f"{diz['Adm']} - {diz['C1']} -> {diz['Corr']}")
    if diz['Corr'] > soglia_corr:
        index += 1
print(f"Totale: {len(c1a)}\nCorrelazioni significative: {index}\nCorrelazioni non significative: {len(c1a) - index}")

admants1_appliances - categories1_homeandgarden -> 0.3089690677434345
admants1_artandentertainment - categories1_artandentertainment -> 0.9601077859472201
admants1_automotive - categories1_automotive -> 0.9674976978644795
admants1_boatinsurance - categories1_travel -> 0.09355831915897822
admants1_businessandindustrialsupplies - categories1_business -> 0.9245505235219128
admants1_careers - categories1_careers -> 0.9582480529903642
admants1_carinsurance - categories1_automotive -> 0.9778606903439595
admants1_consumerelectronics - categories1_technologyandcomputing -> 0.42432316844855866
admants1_dating - categories1_uncategorized -> 0.25334174148424476
admants1_education - categories1_education -> 0.7003388694644928
admants1_energy - categories1_business -> 0.2698033300711038
admants1_familyandparenting - categories1_familyandparenting -> 0.88616628881027
admants1_finance - categories1_finance -> 0.9839754374948233
admants1_foodanddrink - categories1_foodanddrink -> 0.9698346389294276
ad

La maggior parte degli admants ha una correlazione superiore a 0.7 con le categories1.

### Eliminazione 

In [ ]:
df.drop(columns=adms, inplace=True)

## Sentiments

In [ ]:
sents = ['sentiments1_neutroneutralsentiment',
         'sentiments1_sentimentnegativesentiment',
         'sentiments1_sentimentneutralsentiment',
         'sentiments1_sentimentpositivesentiment']

In [ ]:
len(df[df[sents[0]] != 0][sents])

0

In [ ]:
len(df[df[sents[1]] != 0][sents])

0

In [ ]:
len(df[df[sents[2]] != 0][sents])

0

In [ ]:
len(df[df[sents[3]] != 0][sents])

0

Le colonne sono di soli 0. Le elimineremo.

In [ ]:
# Eliminazione

df.drop(columns=sents, inplace=True)

## Feelings

### Normalizzazione

In [ ]:
fel = re.compile('feelings+.*', re.IGNORECASE)
feel = []

for v in df.columns:
  if fel.search(v):
    feel.append(v)
  
# Calcolo somma
def somma_fel(riga):
  sum = 0
  for cat in feel:
    sum += riga[cat]
  return(round(sum, 1))

df['somma'] = df.apply(somma_fel, axis=1)

In [ ]:
# controllo quanti nan ci sono nelle somme: non ci sono nan

df[df['somma'].isnull()][['ad_form_id', 'somma']]

Empty DataFrame
Columns: [ad_form_id, somma]
Index: []

In [ ]:
# controllo se ci sono nan nelle categorie singole: non ci sono

for v in feel:
  nn = len(df[df[v].isnull()])
  if (nn != 0):
    print(v, nn)

In [ ]:
# controllo quanti inf ci sono nelle somme: non ci sono inf
df[df['somma'] == np.inf][['ad_form_id', 'somma']]

Empty DataFrame
Columns: [ad_form_id, somma]
Index: []

In [ ]:
# controllo gli zero: ce ne sono
df[df['somma'] == 0][['ad_form_id', 'somma']]

ad_form_id  somma
1      8408205703932483258    0.0
4      6157419832685345406    0.0
7      6272249560265519045    0.0
8      2248420359615433360    0.0
9      1520530489999888701    0.0
...                    ...    ...
82557  2460917516619769780    0.0
82558  2745741610031721805    0.0
82560  6149323461207714452    0.0
82561  7551278096513860694    0.0
82563  7360059250959897258    0.0

[44610 rows x 2 columns]

Guardiamo come sono i click e se ce ne sono di superiori a zero.

In [ ]:
df[(df['somma'] == 0) & (df['clicks'] != 0)][['ad_form_id', 'clicks', 'somma']]

ad_form_id  clicks  somma
96     6910530872923318739       1    0.0
502    4003952920417380023       1    0.0
1655   1040537780055031487       1    0.0
2311   4900450931963021725       1    0.0
3208   5535943526712202648       1    0.0
...                    ...     ...    ...
79874   161282269580390342       1    0.0
80266   622871437058277142       1    0.0
80411    60933584475798861       1    0.0
81923     8208250892075842       1    0.0
82446  9206113435616205427       1    0.0

[103 rows x 3 columns]

Per evitare di avere normalizzazioni a NaN: poniamo gli zero nelle somme uguali a un valore arbitrario in modo che la divisione non dia NaN.

In [ ]:
soglia = 10

def somma_soglia(riga):
  if riga['somma'] == 0:
    return soglia
  else:
    return riga['somma']
  
df['somma'] = df.apply(somma_soglia, axis=1)

In [ ]:
# Normalizzo le somme a 100
for cat in feel:
  df[cat] = round((df[cat] / df['somma']) * 100, 2)

In [ ]:
df[feel]

feelings1_active  feelings1_affection  feelings1_amusement  \
1                   0.0                  0.0                 0.00   
4                   0.0                  0.0                 0.00   
5                   0.0                  0.0                56.25   
7                   0.0                  0.0                 0.00   
8                   0.0                  0.0                 0.00   
...                 ...                  ...                  ...   
82559               0.0                  0.0                56.10   
82560               0.0                  0.0                 0.00   
82561               0.0                  0.0                 0.00   
82562               0.0                  0.0                56.17   
82563               0.0                  0.0                 0.00   

       feelings1_anger  feelings1_anguish  feelings1_anxiety  \
1                  0.0                0.0                0.0   
4                  0.0                0.0                0.0   
5                  0.0                0.0                0.0   
7                  0.0                0.0                0.0   
8                  0.0                0.0                0.0   
...                ...                ...                ...   
82559              0.0                0.0                0.0   
82560              0.0                0.0                0.0   
82561              0.0                0.0                0.0   
82562              0.0                0.0                0.0   
82563              0.0                0.0                0.0   

       feelings1_behaviour  feelings1_belonging  feelings1_boredom  \
1                      0.0                  0.0                0.0   
4                      0.0                  0.0                0.0   
5                      0.0                  0.0                0.0   
7                      0.0                  0.0                0.0   
8                      0.0                  0.0                0.0   
...                    ...                  ...                ...   
82559                  0.0                  0.0                0.0   
82560                  0.0                  0.0                0.0   
82561                  0.0                  0.0                0.0   
82562                  0.0                  0.0                0.0   
82563                  0.0                  0.0                0.0   

       feelings1_calm  ...  feelings1_surprise  feelings1_sympathy  \
1                 0.0  ...                 0.0                 0.0   
4                 0.0  ...                 0.0                 0.0   
5                 0.0  ...                 0.0                 0.0   
7                 0.0  ...                 0.0                 0.0   
8                 0.0  ...                 0.0                 0.0   
...               ...  ...                 ...                 ...   
82559             0.0  ...                 0.0                 0.0   
82560             0.0  ...                 0.0                 0.0   
82561             0.0  ...                 0.0                 0.0   
82562             0.0  ...                 0.0                 0.0   
82563             0.0  ...                 0.0                 0.0   

       feelings1_thirst  feelings1_thoughtful  feelings1_torment  \
1                   0.0                   0.0                0.0   
4                   0.0                   0.0                0.0   
5                   0.0                   0.0                0.0   
7                   0.0                   0.0                0.0   
8                   0.0                   0.0                0.0   
...                 ...                   ...                ...   
82559               0.0                   0.0                0.0   
82560               0.0                   0.0                0.0   
82561               0.0                   0.0                0.0   
82562               0.0                   0.0  

In [ ]:
df.drop('somma', axis=1, inplace=True)

### Colonne completamente a zero

In [ ]:
index = 0
for fel in feel:
  if len(df[df[fel] == 0]) == len(df):
    print(fel)
    index += 1
print(f'Numero di feelings con entrate tutte 0: {index} / {len(feel)}')

feelings1_anguish
feelings1_categoriauncategorized
feelings1_citylover
feelings1_disillusion
feelings1_emotionsneutralemotions
feelings1_fanatichealthfanatic
feelings1_find
feelings1_intentions
feelings1_lovercitylover
feelings1_lovernaturelover
feelings1_naturelover
feelings1_negativonegative
feelings1_purchase
feelings1_resignation
feelings1_search
feelings1_selfrealization
feelings1_sitelistblindnositelist
feelings1_torment
feelings1_uncategorized
Numero di feelings con entrate tutte 0: 19 / 109


### Distribuzione

In [ ]:
dist_f = [] # vettore di dizionari in cui salvo il p-value del test di student
conteggio_distr_identiche_f = 0
for cat in feel:
    diz = {}

    _, p = scipy.stats.mannwhitneyu(df_y[cat], df_n[cat])

    diz['Feelings'] = cat
    diz['p-value'] = p

    if diz['p-value'] > alfa:
        diz['Esito'] = 'Distribuzione identica'
        conteggio_distr_identiche_f+= 1
    
    else:
        diz['Esito'] = 'Distribuzione diversa'
    
    dist_f.append(diz)

In [ ]:
print(f"Frazione di Feelings con distribuzione identica tra click e non click: {round(conteggio_distr_identiche_f / len(feel), 2)}.")

Frazione di Feelings con distribuzione identica tra click e non click: 0.93.


In [ ]:
for diz in dist_f:
  if diz['Esito'] == 'Distribuzione diversa':
    print(diz)

{'Feelings': 'feelings1_amusement', 'p-value': 5.004242711780167e-11, 'Esito': 'Distribuzione diversa'}
{'Feelings': 'feelings1_confusion', 'p-value': 0.022871942855312827, 'Esito': 'Distribuzione diversa'}
{'Feelings': 'feelings1_emotions', 'p-value': 2.6680903613832493e-08, 'Esito': 'Distribuzione diversa'}
{'Feelings': 'feelings1_forgiveness', 'p-value': 1.1391676239311548e-07, 'Esito': 'Distribuzione diversa'}
{'Feelings': 'feelings1_irritation', 'p-value': 3.036327121079206e-13, 'Esito': 'Distribuzione diversa'}
{'Feelings': 'feelings1_joy', 'p-value': 0.009017388595248657, 'Esito': 'Distribuzione diversa'}
{'Feelings': 'feelings1_pain', 'p-value': 0.006581253847615384, 'Esito': 'Distribuzione diversa'}
{'Feelings': 'feelings1_positive', 'p-value': 1.5038396592723398e-05, 'Esito': 'Distribuzione diversa'}


#### Analisi della frazione di 0

###### Click e non click separati

In [ ]:
zeri_cnc = []
for fel in feel:
  diz = {}
  zeri_y = len(df_y[df_y[fel] == 0])
  zeri_n = len(df_n[df_n[fel] == 0])

  diz['Feeling'] = fel
  diz['% 0 tra i click'] = round(zeri_y / len(df_y), 3)
  diz['% 0 tra i non click'] = round(zeri_n / len(df_n), 3)

  zeri_cnc.append(diz)

In [ ]:
for diz in zeri_cnc:
  print(diz)

{'Feeling': 'feelings1_active', '% 0 tra i click': 1.0, '% 0 tra i non click': 0.998}
{'Feeling': 'feelings1_affection', '% 0 tra i click': 1.0, '% 0 tra i non click': 1.0}
{'Feeling': 'feelings1_amusement', '% 0 tra i click': 0.571, '% 0 tra i non click': 0.77}
{'Feeling': 'feelings1_anger', '% 0 tra i click': 1.0, '% 0 tra i non click': 1.0}
{'Feeling': 'feelings1_anguish', '% 0 tra i click': 1.0, '% 0 tra i non click': 1.0}
{'Feeling': 'feelings1_anxiety', '% 0 tra i click': 1.0, '% 0 tra i non click': 0.999}
{'Feeling': 'feelings1_behaviour', '% 0 tra i click': 0.989, '% 0 tra i non click': 0.98}
{'Feeling': 'feelings1_belonging', '% 0 tra i click': 1.0, '% 0 tra i non click': 1.0}
{'Feeling': 'feelings1_boredom', '% 0 tra i click': 1.0, '% 0 tra i non click': 0.999}
{'Feeling': 'feelings1_calm', '% 0 tra i click': 1.0, '% 0 tra i non click': 1.0}
{'Feeling': 'feelings1_categoriauncategorized', '% 0 tra i click': 1.0, '% 0 tra i non click': 1.0}
{'Feeling': 'feelings1_citylover', '

Verifico se ci sono colonne con % di 0 minore di 0.99.

In [ ]:
for diz in zeri_cnc:
  if diz['% 0 tra i click'] < 0.99 or diz['% 0 tra i non click'] < 0.99:
    print(diz)

{'Feeling': 'feelings1_amusement', '% 0 tra i click': 0.571, '% 0 tra i non click': 0.77}
{'Feeling': 'feelings1_behaviour', '% 0 tra i click': 0.989, '% 0 tra i non click': 0.98}
{'Feeling': 'feelings1_emotions', '% 0 tra i click': 0.545, '% 0 tra i non click': 0.716}
{'Feeling': 'feelings1_motivations', '% 0 tra i click': 0.974, '% 0 tra i non click': 0.979}
{'Feeling': 'feelings1_pain', '% 0 tra i click': 0.989, '% 0 tra i non click': 0.998}
{'Feeling': 'feelings1_power', '% 0 tra i click': 0.989, '% 0 tra i non click': 0.994}
{'Feeling': 'feelings1_sadness', '% 0 tra i click': 0.989, '% 0 tra i non click': 0.996}
{'Feeling': 'feelings1_success', '% 0 tra i click': 0.989, '% 0 tra i non click': 0.994}


##### Click e non click uniti

In [ ]:
zeri_tot = []
for fel in feel:
  diz = {}
  zeri = len(df[df[fel] == 0])

  diz['Feeling'] = fel
  diz['% 0'] = round(zeri / len(df), 3)

  zeri_tot.append(diz)

In [ ]:
for diz in zeri_tot:
  print(diz)

{'Feeling': 'feelings1_active', '% 0': 0.998}
{'Feeling': 'feelings1_affection', '% 0': 1.0}
{'Feeling': 'feelings1_amusement', '% 0': 0.77}
{'Feeling': 'feelings1_anger', '% 0': 1.0}
{'Feeling': 'feelings1_anguish', '% 0': 1.0}
{'Feeling': 'feelings1_anxiety', '% 0': 0.999}
{'Feeling': 'feelings1_behaviour', '% 0': 0.98}
{'Feeling': 'feelings1_belonging', '% 0': 1.0}
{'Feeling': 'feelings1_boredom', '% 0': 0.999}
{'Feeling': 'feelings1_calm', '% 0': 1.0}
{'Feeling': 'feelings1_categoriauncategorized', '% 0': 1.0}
{'Feeling': 'feelings1_citylover', '% 0': 1.0}
{'Feeling': 'feelings1_committed', '% 0': 1.0}
{'Feeling': 'feelings1_compassionate', '% 0': 1.0}
{'Feeling': 'feelings1_conformist', '% 0': 1.0}
{'Feeling': 'feelings1_confusion', '% 0': 0.999}
{'Feeling': 'feelings1_courage', '% 0': 0.998}
{'Feeling': 'feelings1_curiosity', '% 0': 0.997}
{'Feeling': 'feelings1_desire', '% 0': 0.997}
{'Feeling': 'feelings1_disappointment', '% 0': 1.0}
{'Feeling': 'feelings1_disgust', '% 0': 1.0}

Verifico se ci sono colonne con % di 0 minore di 0.99.

In [ ]:
for diz in zeri_tot:
  if diz['% 0'] < 0.99:
    print(diz)

{'Feeling': 'feelings1_amusement', '% 0': 0.77}
{'Feeling': 'feelings1_behaviour', '% 0': 0.98}
{'Feeling': 'feelings1_emotions', '% 0': 0.716}
{'Feeling': 'feelings1_motivations', '% 0': 0.979}


##### Distribuzione click e non click nelle 8 colonne notevoli

Per le colonne con almeno una tra la frazione di zeri per i click e la frazione di zeri per i non click minori di 0.99 visualizzo i risultati del test sulla distribuzione

In [ ]:
for diz in dist_f:
  if diz['Feelings'] in ['feelings1_amusement','feelings1_behaviour','feelings1_emotions','feelings1_motivations','feelings1_pain','feelings1_power', 'feelings1_sadness','feelings1_success']:
    print(diz)

{'Feelings': 'feelings1_amusement', 'p-value': 5.004242711780167e-11, 'Esito': 'Distribuzione diversa'}
{'Feelings': 'feelings1_behaviour', 'p-value': 0.3650813126870386, 'Esito': 'Distribuzione identica'}
{'Feelings': 'feelings1_emotions', 'p-value': 2.6680903613832493e-08, 'Esito': 'Distribuzione diversa'}
{'Feelings': 'feelings1_motivations', 'p-value': 0.6276914311975152, 'Esito': 'Distribuzione identica'}
{'Feelings': 'feelings1_pain', 'p-value': 0.006581253847615384, 'Esito': 'Distribuzione diversa'}
{'Feelings': 'feelings1_power', 'p-value': 0.3825854221851541, 'Esito': 'Distribuzione identica'}
{'Feelings': 'feelings1_sadness', 'p-value': 0.17245428431323173, 'Esito': 'Distribuzione identica'}
{'Feelings': 'feelings1_success', 'p-value': 0.36553993620266945, 'Esito': 'Distribuzione identica'}


##### Frazione di zeri click e non click separati nelle colonne con distribuzione diversa

Per le colonne con distribuzioni diverse tra click e non click valuto la frazione di zeri tra click e non click

In [ ]:
for i in range(len(dist_f)):
  if dist_f[i]['Esito'] == 'Distribuzione diversa':
    print(zeri_cnc[i])

{'Feeling': 'feelings1_amusement', '% 0 tra i click': 0.571, '% 0 tra i non click': 0.77}
{'Feeling': 'feelings1_confusion', '% 0 tra i click': 0.995, '% 0 tra i non click': 0.999}
{'Feeling': 'feelings1_emotions', '% 0 tra i click': 0.545, '% 0 tra i non click': 0.716}
{'Feeling': 'feelings1_forgiveness', '% 0 tra i click': 0.995, '% 0 tra i non click': 1.0}
{'Feeling': 'feelings1_irritation', '% 0 tra i click': 0.995, '% 0 tra i non click': 1.0}
{'Feeling': 'feelings1_joy', '% 0 tra i click': 0.995, '% 0 tra i non click': 0.999}
{'Feeling': 'feelings1_pain', '% 0 tra i click': 0.989, '% 0 tra i non click': 0.998}
{'Feeling': 'feelings1_positive', '% 0 tra i click': 0.995, '% 0 tra i non click': 1.0}


##### feelings_others

Vengono mantenute le colonne che presentano contemporaneamente:
* almeno una tra la frazione di zeri per i click e la frazione di zeri per i non click minore di 0.99 (per i sì circa 187 su 189, per i no circa§ 61568 su 62190)
* distribuzione diversa tra click e non click.

Le altre colonne sono accorpate in others.

In [ ]:
def feeling_others(riga):
  sum = 0
  for fel in feel:
    if fel not in ['feelings1_amusement','feelings1_emotions','feelings1_pain']:
      sum += riga[fel]
  
  return sum

df['feelings1_others'] = df.apply(feeling_others, axis=1)

In [ ]:
df[['feelings1_others','feelings1_amusement','feelings1_emotions','feelings1_pain']]

feelings1_others  feelings1_amusement  feelings1_emotions  \
1                   0.0                 0.00                0.00   
4                   0.0                 0.00                0.00   
5                   0.0                56.25               43.75   
7                   0.0                 0.00                0.00   
8                   0.0                 0.00                0.00   
...                 ...                  ...                 ...   
82559               0.0                56.10               43.90   
82560               0.0                 0.00                0.00   
82561               0.0                 0.00                0.00   
82562               0.0                56.17               43.96   
82563               0.0                 0.00                0.00   

       feelings1_pain  
1                 0.0  
4                 0.0  
5                 0.0  
7                 0.0  
8                 0.0  
...               ...  
82559             0.0  
82560             0.0  
82561             0.0  
82562             0.0  
82563             0.0  

[62379 rows x 4 columns]

In [ ]:
df[(df['feelings1_others'] == 0) & (df['feelings1_amusement'] == 0) & (df['feelings1_emotions'] == 0) & (df['feelings1_pain'] == 0)][['feelings1_others','feelings1_amusement','feelings1_emotions', 'feelings1_pain']]

feelings1_others  feelings1_amusement  feelings1_emotions  \
1                   0.0                  0.0                 0.0   
4                   0.0                  0.0                 0.0   
7                   0.0                  0.0                 0.0   
8                   0.0                  0.0                 0.0   
9                   0.0                  0.0                 0.0   
...                 ...                  ...                 ...   
82557               0.0                  0.0                 0.0   
82558               0.0                  0.0                 0.0   
82560               0.0                  0.0                 0.0   
82561               0.0                  0.0                 0.0   
82563               0.0                  0.0                 0.0   

       feelings1_pain  
1                 0.0  
4                 0.0  
7                 0.0  
8                 0.0  
9                 0.0  
...               ...  
82557             0.0  
82558             0.0  
82560             0.0  
82561             0.0  
82563             0.0  

[44610 rows x 4 columns]

Ci sono righe che hanno entrate pari a 0 in tutte e 4 le colonne (44610) e questo non dovrebbe essere possibile (mancanza di informazione). Guardo se ci sono entrate con others diverso da 0.

In [ ]:
df[df['feelings1_others'] > 0][['feelings1_others','feelings1_amusement','feelings1_emotions','feelings1_pain']]

feelings1_others  feelings1_amusement  feelings1_emotions  \
16                58.33                 0.00               41.67   
26                48.78                29.27               21.95   
29                69.43                 0.00               30.63   
40                69.49                 0.00               30.66   
42                 5.61                50.49               43.90   
...                 ...                  ...                 ...   
82484             76.92                 0.00               23.08   
82492             55.56                 0.00               44.44   
82493             48.78                29.27               21.95   
82498             48.84                29.30               21.98   
82530             72.80                 0.00               26.82   

       feelings1_pain  
16                0.0  
26                0.0  
29                0.0  
40                0.0  
42                0.0  
...               ...  
82484             0.0  
82492             0.0  
82493             0.0  
82498             0.0  
82530             0.0  

[4135 rows x 4 columns]

Ci sono entrate con others diverso da 0 (4135).

La somma non è proprio 100, ma è per via dei round applicati in fase di normalizzazione. 

In [ ]:
# elimino tutti i feelings1 a parte i 4 sopra
feel2 = [] # feelings1 da elimninare
for fel in feel:
  if fel not in ['feelings1_others','feelings1_amusement','feelings1_emotions','feelings1_pain']:
    feel2.append(fel)
len(feel2)

106

In [ ]:
df.drop(columns=feel2, inplace=True)

# Nuovo dataset (dataset_normalizzato)





In [ ]:
colonne = list(df.columns)
for colonna in colonne:
  print(colonna)

ad_form_id
clicks
impressions
os_android
os_bsd
os_ios
os_linux
os_osx
os_other
os_windows
device_type
browser_android
browser_chrome
browser_chromium
browser_edge
browser_firefox
browser_ie
browser_opera
browser_other
browser_safari
browser_unknown
time1_workday_morning
time1_workday_afternoon
time1_workday_evening
time1_workday_night
time1_weekend_morning
time1_weekend_afternoon
time1_weekend_evening
time1_weekend_night
L00_50
L51_100
L101_250
L251_500
L501_1000
L1001_2500
L2501_5000
L5001_10000
L10001_more
categories1_artandentertainment
categories1_automotive
categories1_business
categories1_careers
categories1_education
categories1_emotions
categories1_familyandparenting
categories1_finance
categories1_foodanddrink
categories1_healthandfitness
categories1_hobbiesandinterests
categories1_homeandgarden
categories1_intentions
categories1_lawgovtandpolitics
categories1_news
categories1_pets
categories1_realestate
categories1_religionandspirituality
categories1_science
categories1_shop

In [ ]:
# Export in csv
# pd.DataFrame.to_csv(df.set_index('ad_form_id'), '/content/gdrive/My Drive/DS Lab - PROGETTO/Ottobre/dataset_normalizzato.csv')